### DISCUSSION TITLE
[3.10] Let's move ahead from alpha to beta

### DISCUSSION THREAD

**Discussion by sozzled** [joomla/joomla-cms#34468] [General] [answered]
As far as I'm concerned (being one person among a dozen others who has trialed J! 3.10), I think we should move the status of J! 3.10 alpha into J! 3.10 beta (or perhaps even into RC status).

I had an old J! 3.9 test website.  Nothing extravagant:  a few user accounts, articles, images, with _Akeeba Backup_ installed; the default **Protostar** template.  It's a fairly typical example of a J! website that anyone would be familiar with.  It's "bread and butter" Joomla.

I updated this website to J! 3.10-alpha6 yesterday.  No problems with the update; it happened in a matter of a few minutes.  After updating the website, I added another user, ran _Akeeba Backup_, added a couple more articles, uploaded another image or two.  Again, fairly normal operations.  I've spent about four hours testing this website with _Mozilla Firefox_, _Opera_, _Google Chrome_, _MS Edge_, _Safari_ (for iPad) and _Safari_ (for iPhone); frontend and backend.  I've installed two new extensions (using both the **Install from Web** and **Upload Package File** variants) , enabled them, disabled them, uninstalled them.  I've tested the **Search** component.

I ran the **Pre-Update Check**; it said there were "potential serious issues" migrating the site to J! 4 (if I were to proceed) with _Akeeba Backup_.  I don't think this is a show-stopper because I've used the same version of _Akeeba Backup_ on J! 4 RC1 and it works quite well.  The **Pre-Update Check** displays fine on my browser.

I've played with the **Redirect** component, progressive [server-side] caching, forced HTTPS (front/backend/both), database vs. PHP session handling.   Everything works here.

I haven't found any problems.  How many more things should I test?  The J! 3.10 [alpha6] site works as well as J! 3.9.27 on which it was based.  What more do we need to do?

I can't see anything that prevents J! 3.10 moving into the beta phase.  I think it would be good if J! 3.10 moved into beta because it signals the J! community that J! 3.10 development has progressed and we're closer to completing this part of the project.

**Comment by richard67** [comment]
Ping @zero-24 .

**Comment by zero-24** [comment]
Hi,

thanks for testing, its great to get positive feedback like this.

The only reason 3.10 is still alpha is that 3.10 is not feature complete yet but waits for the EOS plugin to be finalised. Once that plugin can be merged in an initail version the beta / RC is near :-)

**Reply by sozzled** [reply]
How much longer will it take to write the [EOS plugin](https://issues.joomla.org/tracker/joomla-cms/34453)?

Seriously, I think that eight months of "alpha testing" is too long. 

I know that a "decision" was made to hold back the release of J! 3.10 stable until J! 4.0 stable is released but it looks like this is a marketing decision and not a technical decision.  I think that J! 3.9._x_ should be mothballed and release J! 3.10 as the final in the J! 3._x_ series ASAP.  Holding back J! 3.10 now looks crazy ... not to mention that these hold-ups are turning people away from the project.

As far as I'm concerned, even without a few lines of EOS plugin that should take about an hour to write, J! 3.10 is ready to move forward at a much accelerated pace.  We could even see J! 3.10 stable released in two months' time, couldn't we, even **_without_** J! 4.0 stable?

**Reply by zero-24** [reply]
The EOS Plugin is already in testing and finalisation phase so i dont expect it to take that long to complet it.

The main reason to "hold back" 3.10 Is that two years of support right after the final release and that 3.10 is going to be the base for 4.0 upgrades.

Other than the pre upgrade checker, which is mostly relevant to 4.x upgrades, and the EOS plugin nothing major has been changed between 3.10 and 3.9. so releasing it now would not bring any benifit for existing sites other than an increased version number and a reduced time this version will be supported.

Given that 3.10 and 4.0 have always been announced to be released together and regulary simultan to 4.x releases I dont see why we should change that.

The main focus should (and already is) on getting 4.0 ready so we can release the first stable as early as possible, together with a stable release of 3.10.

**Reply by brianteeman** [reply]
> even without a few lines of EOS plugin that should take about an hour to write,

I 've spent several hours on it already just debugging it

**Reply by sozzled** [reply]
> Given that 3.10 and 4.0 have always been announced to be released together and regulary simultan to 4.x releases I dont see why we should change that.

I disagree.  I think that the decision should be revisited in light of what I wrote earlier.  Until J! 3.10 is released, J! 3.9._x_ will continue to be updated with various patches (that also need to be merged with J! 3.10) and signals a lack of progress moving towards the end of the J! 3._x_ series development.  By moving into J! 3.10 now (or after successfully testing beta/RC) we demonstrate that the project is moving ahead and we encourage more people to update to J! 3.10 sooner rather than later.

This also shows that progress is being made towards completing J! 4.  Remember that J! 4 has been six years in the making and the lack of progress (in terms of finalising J! 3._x_ or releasing J! 4) is turning users away from Joomla!

I refer you to the announcement made four years ago:  https://developer.joomla.org/news/676-joomla-3-retiring-as-joomla-4-comes-to-life.html.  Four years later and J! 3.x shows no sign of being retired in the foreseeable future.  If we're going to be serious about retiring J! 3._x_ then I believe we should get J! 3.10 out.   Further, the sooner that J! 3.10 is released, the more likely it will be that people will return to J! and become interested in the project.

I would like people to look again at the decision to simultaneously release J! 3.10 with J! 4.0.  Aside from the mixed message that this simultaneous release strategy sends, it's confusing.  It sends the message, "Should I spend the next two years planning my migration to J! 4, should I update to J! 3.10 or should I just start all over—trash my existing J! 3.x websites—and begin fresh with J! 4?"  I believe those will be the main questions that people will ask, and their questions will become more frequent if we delay further with J! 3.10.

Thanks, @brianteeman for debugging the EOS plugin.  I would have thought, because similar plugins were developed when J! 1.6 and J! 3.0 were released that the principles should be reasonably well established, that the effort in writing an EOS plugin shouldn't be too difficult.  But, then again, since the last time an EOS plugin was written (12 years ago or so), maybe we don't have that knowledge?  I don't want to criticise the effort in writing the plugin but, perhaps is it really a show-stopper to releasing J! 3.10 beta (or J! 3.10 RC1)?

> The main focus should (and already is) on getting 4.0 ready so we can release the first stable as early as possible, together with a stable release of 3.10.

I think the main focus should be on motivating people to use J! (irrespective of whatever version it may be).  Certainly it is desirable, from a support viewpoint, to have people using the latest version.  It's disappointing to see people asking about J! 2.5 or earlier on the J! forum.  But, most importantly, we've had little progress to show for J! since FftF and people are right to feel concerned about the direction that J! is heading.

**Comment by brianteeman** [comment]
J3.10 will be released at the same time as J4.0 - fullstop

Doesn't matter how many hundreds of words that you write about everyone not working hard enough or not being skilled enough.

**Reply by sozzled** [reply]
The decision, that it is a necessary condition for J! 4.0 stable to be available for releasing J! 3.10 stable, was made (no doubt with the best of intentions) at least three years ago:  see https://web.archive.org/web/20180623151651/https://developer.joomla.org/roadmap.html

That announcement took place before FftF.  Since the announcement was made in 2018 the world changed—GDPR threw a spanner in the works, for example, and plans changed—and there have been thirty-eight releases in J! 3.9._x_.  I suspect there will be further updates to J! 3.9._x_. between now and the release of J! 3.10 and/or J! 4.  I don't really think that saying "full stop" or end-of-story advances the conversation constructively.

Can someone please provide me good reasons why it's a necessary condition for J! 4 to be released in order to release J! 3.10?   Why not?

To put it another way, what happens if J! 3.10 is released?  Wouldn't this encourage people to update?  And, on updating, people would be informed about what J! 3.10 means as far as the future of their websites is concerned?

That's not "hundreds of words".  That's just asking for a review of the release strategy, that's all.

Decisions can **_always_** be reviewed taking into account current circumstances.  :wink: 

I never wrote **_anywhere_** that people (or everyone) are not working "hard enough".  If I wrote that **_anywhere_** then please show me.  I respect the diligence of people's work.  I'm working as hard as I can to help.  I should like to think that you might look carefully at what I've written—in the spirit of assisting the project—as opposed to characterising what I've written as if I'm approaching things negatively.

I've tested J! 3.10-alpha6 as far as I can for the time being.  It works.  Things are looking good.  I'm looking forward to what happens next. I don't know how more positive I can be that that.  :smiley:

**Comment by zero-24** [comment]
> To put it another way, what happens if J! 3.10 is released? Wouldn't this encourage people to update? And, on updating, people would be informed about what J! 3.10 means as far as the future of their websites is concerned?

When 3.10 and 4.0 are released 3.10 is an update to be installed on all 3.9. sites, after that you have two years to plan and prepare the migration or new install with 4. On 3.10 (as mention above) there will be a EOS plugin informing the existing sites on 4.x as well as direct to migration guides. And that 3.10 goes out of support after two years has been annouced [many](https://developer.joomla.org/news/764-joomla-3-10-and-joomla-4-0.html) [times](https://developer.joomla.org/news/764-joomla-3-10-and-joomla-4-0.html) and no changes to that are planed. :)

> Decisions can always be reviewed taking into account current circumstances. 😉

Sure and sometimes it makes sense to stay by that what you have said so others can see that they can rely on what you say :)
https://developer.joomla.org/news/764-joomla-3-10-and-joomla-4-0.html
https://magazine.joomla.org/all-issues/november-2020/joomla-3-10-status-release-strategy

> I've tested J! 3.10-alpha6 as far as I can for the time being. It works. Things are looking good. I'm looking forward to what happens next. I don't know how more positive I can be that that. 😃

Thats good and I'm happy to hear that. Once the EOS plugin is ready the next step is a RC version of 3.10 awaiting the final release timing for 4. :)

Thanks.

**Reply by sozzled** [reply]
Thanks, Tobias.  You will recall my comment on your magazine article—thanks again for your article.

I wrote
>Thanks, Tobias, for recapping the situation with J! 3.10 and we thank you for keeping us informed. As much as I (among others who are less savvy about GitHub as the developers are) am happy to lend my assistance in testing a pre-stable release of J! 3.10, I am not an alpha tester. I made the decision to not be an alpha tester for anything many years ago and I live in blissful ignorance of the intricate technicalities involved in understanding, building, testing and bug-fixing early versions of software particularly when end-user documentation has not yet been written. Being an alpha tester means, in large part, taking responsibility for one's own actions and being prepared to submit to the decisions of others who may not fully appreciate the views or opinions of other alpha testers in arriving at the point when the feature set(s) that these alpha versions will or will not include.

Likewise, I made a decision years ago to not test alpha releases of software.  I don't have the skills to debug and fix problems with alpha software.  I could report bugs, I suppose, but alpha software is volatile and rapidly changes from day to day.

I reviewed my decision (about not being an alpha tester) and decided to give J! 3.10-alpha6 a trial.  I would like to suggest, with the great respect that I have for your work, that you bring this discussion to the next team/department meeting and have them look at it.  That can't hurt to do that.  If the decision is reaffirmed then that's OK; if the team decides to change the strategy—in order to demonstrate to ordinary users that J! 3.10 is a good idea—then that's good, too. :smile:

**Comment by sozzled** [comment]
Updated to J! 3.10-alpha7 a short time ago.  I don't see any operational difference between this release and the previous one.

**Reply by zero-24** [reply]
Yes nothing major changed.

**Reply by sozzled** [reply]
Thanks, Tobias.  Are we any closer to a **_beta_** release?  Hows the EOS plugin going?  Will you put the idea on the table to the CMS Release Team to reconsider the decision about a simultaneous release of J! 3.10 with J! 4.0?  I can't see (from a technical perspective) why these two matters are coupled together and it may be better, to encourage people to move towards J! 4 if J! 3.10 was released beforehand.  This would give people the additional motivation to climb back on board the J! project; our concerns are that people have lost confidence that anything new will happen in the very near future.  I think that releasing J! 3.10 sooner will help restore some of that confidence.

**Reply by zero-24** [reply]
Nothing has been changed in the release plans. As for the EOS Plugin you can follow the PR regarding that. The last thing we are waiting for are the final strings for the messages.

Once the strings are there as well at the PR got tested and meged the next update will be a RC version until that there will be alpha releases released together with 4.0-rc 's

**Reply by sozzled** [reply]
Thanks, Tobias.  From a project management perspective, this is somewhat disappointing news.

Currently there are three development streams running separately and in parallel:  (a) J! 3.9, (b) J! 3.10 and (c) J! 4.0.  I suppose you could say there are four development streams running, if you add J! 4.1.

From an overall project management viewpoint, making sure that all the of new developments in J! 3.9 find their way into J! 3.10 is a problem that has to be managed.  As long as these two development streams are working collaboratively and in lock-step with one another then that's OK.  My understanding is that when new things happen in J! 3.9 it's your job to ensure those things are incorporated in J! 3.10.  The longer that J! 3.9 continues to be the stable, supported version the longer you have to "play catch-up".

The same situation applies with J! 4.0 and J! 4.1.

At some point, J! 4.0 is released and the whole dynamic changes.  People then choose between staying with J! 3.x or adopting J! 4._x_; the third choice that people migrate from J! 3._x_ to J! 4.0 (and then update later to subsequent releases) assuming that a migration pathway is available.

The longer this goes on the more likely people will opt-out from taking on new versions of J! and they'll remain wherever they are.

**Reply by zero-24** [reply]
The main development focus is on 4.0. updating 3.10 with the minimal changes done in 3.9 is usually as easy as running like 3 commands on the cli. Done by the release leads like Harald, George or me. So nothing that critical and already managed. :-)

**Reply by zero-24** [reply]
> At some point, J! 4.0 is released and the whole dynamic changes. People then choose between staying with J! 3.x or adopting J! 4.x; the third choice that people migrate from J! 3.x to J! 4.0 (and then update later to subsequent releases) assuming that a migration pathway is available.

Yes thats the intended situation. 4.0.0 will not be as stable as 3.10.0 will because major changes have been made in 4.0. also not all extension will be aviable for 4.0 right from the beginning. Thats the reason that migration periode of 2 years has been implemented. So everyone can choose the best path for thier project.

3.10 -> 4.0 has a migration path for the core system and once the extensions and templates are joomla 4 ready that migration is in the end usually a one-click one

**Comment by sozzled** [comment]
Perhaps I can explain the project management problem more simply like this:  

1)  Changes occur—independently of J! 3.10 and J! 4.x—in J! 3.9.
2)  Every time a new PR is accepted into J! 3.9 those changes need to be added to J! 3.10
3)  Some of these issues may impact on the development of J! 4.0; some of these issues may affect J! 4.1.
4)  If further development streams are added, the development chain grows.  Changes propagate along the chain, from J! 3.9 to J! 3.10 and (perhaps) into J! 4.0/4.1 ... etc.

For as long as J! 3.9._x_ continues to remain the basis of what is "stable" and "supported" the changes introduced into the base need to be factored into other development streams.  I think that summarises the project management problem.

I'm only asking that the CMS Release Team considers the possibility that these changes aren't properly reflected in the various development streams—so that each development stream stays in step with the base—and looks again at the release strategy.  It's possible that the CMS Release Team will re-affirm the release strategy (and what that entails for each development team going forward).  It's possible that the CMS Release Team may change the strategy:  stop further release of J! 3.9 in favour of releasing J! 3.10 now

You don't know unless you ask the question. :smile:

**Reply by brianteeman** [reply]
the reality doesnt match your assumptions as can be seen by how few pr have been applied upstream in the last 6 months

**Reply by sozzled** [reply]
@brianteeman:  Does it matter if there's one accepted PR or more than one?  Any change, regardless of the number of them, affects the other development streams, doesn't it?

Another reality is that there are few people interested in testing alpha releases as can be seen by the number of people engaged in this discussion (or elsewhere, e.g. The Joomla Forum™).

**Comment by sozzled** [comment]
Updated to J! 3.10-alpha8 a short time ago. I don't see any operational difference between this release and the previous two.

**Reply by brianteeman** [reply]
You wont as there havent been any

https://github.com/joomla/joomla-cms/commits/3.10-dev

**Comment by sozzled** [comment]
Updated to J! 3.10-alpha9 a short time ago. I don't see any operational difference between this release and the previous **_three_** releases.   I'm looking for the EOS plugin.  It doesn't seem to exist.

**Reply by zero-24** [reply]
Yes as you can see here the latest status about the EOS Plugin [here](https://github.com/joomla/joomla-cms/pull/34453) still awaiting the final strings to be provided. Once thats done it can be merged and included in the next release.

**Reply by sozzled** [reply]
So what are you saying?  J! 3.10 alpha9 is no different to J! 3.10 alpha8 (which was no different to J! 3.10 alpha7 or J! 3.10 alpha6)?  Why create new releases, have people spend their time installing them, if there's no difference?  Sounds like a Claytons release to me (and a waste of people's time).

**Reply by zero-24** [reply]
There where always changes beeing the than applied changes to the staging branch you can review them here: https://github.com/joomla/joomla-cms/commits/3.10-dev Nothing major changed to be fair.

**Reply by richard67** [reply]
As always,  new alpha contains also latest changes from 3.9 having been merged up, and this includes the security fixes. So saying there has not been any change between the alphas is simply wrong. (just my 5 ct)

**Reply by brianteeman** [reply]
there really wasnt any need or point in alpha 8 or 9

**Reply by sozzled** [reply]
So ... the new release changed the wording "Potentially serious problem" (discussed at https://github.com/joomla/joomla-cms/issues/34599) to "Potential update problem".  That was it.

I agree with @brianteeman.

**Reply by brianteeman** [reply]
can I frame that

**Reply by zero-24** [reply]
> So ... the new release changed the wording "Potentially serious problem" (discussed at #34599) to "Potential update problem". That was it.
> 
> I agree with @brianteeman.

As well as updating to the latest changes done in staging which fixed for example also [security issues](https://github.com/joomla/joomla-cms/commit/87ed5678dbb3d1a3829d638dff53653d88e07513)?

Again yes there are no major updates included cause the main development focus for good reasons is on 4.0.

**Reply by sozzled** [reply]
> Again yes there are no major updates included cause the main development focus for good reasons is on 4.0.

So, if the main development focus is "for good reasons" on J! 4.0 (and, by that, I infer that J! 3.10 is only a sideshow) we might as well cease testing J! 3.10 and put it on hold?

**Reply by zero-24** [reply]
Up to you to decided I would recomend extension developers to test with the latest 3.10 for example the pre upgrade checker as it seems we also just found an edge case that we dint found while testing the initial PR. https://github.com/joomla/joomla-cms/pull/34776. As for the user tests there is nothing major to be tested right now and those testers can go and test 4.0 for now.

**Reply by sozzled** [reply]
Oh good grief.  If there's nothing "major" to be tested right now why bother releasing new releases?  I have asked several times in this discussion for J! 3.10 to move from alpha to beta (and beyond that to RC or even stable) but, for a variety of "good reasons" I have been informed (or advised) that I'm the only person in the entire universe who wants to see J! 3.10 evolve.  We will have to wait until after J! 4 is released ... and I haven't decided whether there's a business future in J! 4 for me.

I'm trying to help.  Seriously, I'm just trying to help.  There may be technical issues such as putting a team of word-smiths to work and put pen to paper and figure out how to succinctly express the EOS notice and I understand that.  There are reputational problems to the J! project by having J! 3.9 continue with no end in sight to releases in future.

**Comment by sozzled** [comment, marked as answer]
Excellent work:  [updating the J! 3.10 roadmap](https://developer.joomla.org/roadmap.html).  A couple more weeks and we'll move into RC. :smile:

**Comment by Hackwar** [comment]
Since Joomla 3 is now EOL, I'm closing this discussion.
